File: antia_numerical_methods/c/fun/ylm_x.c
/*	To compute spherical harmonic Y_lm(COS(THETA),PHI)
	Version of YLM with argument as COS(THETA) instead of THETA

	L : (input) Degree of spherical harmonic
	M : (input) Azimuthal order of spherical harmonic
	X, PHI : (input) Real variables specifying the values of cos(theta)
		and phi at which the spherical harmonic needs to be evaluated
	Y : (output) Array of length 2 containing the complex value of
		spherical harmonic.

	Required functions : PLM
*/

#include <math.h>
#include <stdlib.h>

#define PI 3.14159265358979324

void plm(int l, int m, double x, double p[]);

void ylm_x(int l, int m, double x, double phi, double y[])

{
	int i,mm;
	double clm;
	double *p;

	y[0]=0.0; y[1]=0.0;
	if(l<0 || abs(m)>l) return;

	p=(double *) calloc((size_t) (l+2), sizeof(double));
	mm=abs(m);
 
/*	To use THETA instead of THETA in argument uncomment  this line 
	x=cos(theta); */
	plm(l,mm,x,p);

	clm=(2*l+1.0)/(4.*PI);
	for(i=l-mm+1; i<=l+mm; ++i) clm=clm/i;
	clm=sqrt(clm);

	if(mm - 2*(mm/2) == 1 && m>=0) clm=-clm;
	y[0]=clm*p[l]*cos(m*phi);
	y[1]=clm*p[l]*sin(m*phi);
	free(p);
	return;
}

File: projects/antia_numerical_methods@Beliavsky/f77/sub4/ylm_x.f
!	To compute spherical harmonic Y_lm(COS(THETA),PHI)
!	Version of YLM with argument as COS(THETA) instead of THETA
!
!	L : (input) Degree of spherical harmonic
!	M : (input) Azimuthal order of spherical harmonic
!	X, PHI : (input) Real variables specifying the values of COS(THETA)
!		and PHI at which the spherical harmonic needs to be evaluated
!	YLM_X is complex value of spherical harmonic and must be declared
!		to be complex in the calling routine
!
!	Required routines : PLM
 
      FUNCTION YLM_X(L,M,X,PHI)
      IMPLICIT REAL*4(A-H,O-X)
      IMPLICIT COMPLEX*8(Y,Z)
      PARAMETER(LMAX=5001,PI=3.14159265358979324D0)
      DIMENSION P(LMAX)
 
      YLM_X=0.0
      IF(L.LT.0.OR.ABS(M).GT.L.OR.L.GE.LMAX) RETURN
      MM=ABS(M)
!	To use THETA instead of X in argument uncomment  this line
!      X=COS(THETA)
      CALL PLM(L,MM,X,P)

      CLM=(2*L+1.)/(4.*PI)
      DO 200 I=L-MM+1,L+MM
        CLM=CLM/I
200   CONTINUE
      CLM=SQRT(CLM)

      IF(MOD(MM,2).EQ.1.AND.M.GE.0) CLM=-CLM
      YI=(0.D0,1.D0)
      YLM_X=CLM*P(L+1)*EXP(YI*M*PHI)
      END
